net-layer/third/elfutils/doc/debuginfod-find.1
'\"! tbl | nroff \-man
'\" t macro stdmacro

.de SAMPLE
.br
.RS 0
.nf
.nh
..
.de ESAMPLE
.hy
.fi
.RE
..

.TH DEBUGINFOD-FIND 1
.SH NAME
debuginfod-find \- request debuginfo-related data

.SH SYNOPSIS
.B debuginfod-find [\fIOPTION\fP]... debuginfo \fIBUILDID\fP
.br
.B debuginfod-find [\fIOPTION\fP]... debuginfo \fIPATH\fP
.br
.B debuginfod-find [\fIOPTION\fP]... executable \fIBUILDID\fP
.br
.B debuginfod-find [\fIOPTION\fP]... executable \fIPATH\fP
.br
.B debuginfod-find [\fIOPTION\fP]... source \fIBUILDID\fP \fI/FILENAME\fP
.br
.B debuginfod-find [\fIOPTION\fP]... source \fIPATH\fP \fI/FILENAME\fP
.br
.B debuginfod-find [\fIOPTION\fP]... metadata \fIKEY\fP \fIVALUE\fP

.SH DESCRIPTION
\fBdebuginfod-find\fP queries one or more \fBdebuginfod\fP servers for
debuginfo-related data.  In case of a match, it saves the the
requested file into a local cache, prints the file name to standard
output, and exits with a success status of 0.  In case of any error,
it exits with a failure status and an error message to standard error.

.\" Much of the following text is duplicated with debuginfod.8

The debuginfod system uses buildids to identify debuginfo-related data.
These are stored as binary notes in ELF/DWARF files, and are
represented as lowercase hexadecimal.  For example, for a program
/bin/ls, look at the ELF note GNU_BUILD_ID:

.SAMPLE
% readelf -n /bin/ls | grep -A4 build.id
Note section [ 4] '.note.gnu.buildid' of 36 bytes at offset 0x340:
Owner          Data size  Type
GNU                   20  GNU_BUILD_ID
Build ID: 8713b9c3fb8a720137a4a08b325905c7aaf8429d
.ESAMPLE

Then the hexadecimal BUILDID is simply:

.SAMPLE
8713b9c3fb8a720137a4a08b325905c7aaf8429d
.ESAMPLE

In place of the hexadecimal \fIBUILDID\fP, debuginfod-find also
accepts a \fIPATH\fP name to to an ELF binary, from which it extracts
the buildid.  This alternative is triggered if the file name has some
character other than \fB[0-9a-f]\fP.  Files ambiguously named files
like "\fBdeadbeef\fP" can be passed with a \fB./deadbeef\fP extra path
component.


.SS debuginfo \fIBUILDID\fP

If the given buildid is known to a server, this request will result
in a binary object that contains the customary \fB.*debug_*\fP
sections.  This may be a split debuginfo file as created by
\fBstrip\fP, or it may be an original unstripped executable.

.SS executable \fIBUILDID\fP

If the given buildid is known to the server, this request will result
in a binary object that contains the normal executable segments.  This
may be a executable stripped by \fBstrip\fP, or it may be an original
unstripped executable.  \fBET_DYN\fP shared libraries are considered
to be a type of executable.

.SS source \fIBUILDID\fP \fI/SOURCE/FILE\fP

If the given buildid is known to the server, this request will result
in a binary object that contains the source file mentioned.  The path
should be absolute.  Relative path names commonly appear in the DWARF
file's source directory, but these paths are relative to
individual compilation unit AT_comp_dir paths, and yet an executable
is made up of multiple CUs.  Therefore, to disambiguate, debuginfod
expects source queries to prefix relative path names with the CU
compilation-directory, followed by a mandatory "/".

Note: for software packaged by distributions, the CU
compilation-directory may not be obvious.  It can be found by
inspecting AT_comp_dir values in downloaded debuginfo.  For example,
the comp_dir of the Fedora 37 version of /bin/ls can be found as
follows:

.SAMPLE
% debuginfod-find debuginfo /bin/ls
~/.cache/debuginfod_client/03529d48345409576cd5c82a56ad08555088d353/
% eu-readelf -w ~/.cache/debuginfod_client/03529d48345409576cd5c82a56ad08555088d353/debuginfo | grep comp_dir
           comp_dir             (line_strp) "/usr/src/debug/coreutils-9.1-6.fc37.x86_64/separate"
.ESAMPLE

Note: the caller may or may not elide \fB../\fP or \fB/./\fP or extraneous
\fB///\fP sorts of path components in the directory names.  debuginfod
accepts both forms.  Specifically, debuginfod canonicalizes path names
according to RFC3986 section 5.2.4 (Remove Dot Segments), plus reducing
any \fB//\fP to \fB/\fP in the path.

For example:
.TS
l l.
#include <stdio.h>	source BUILDID /usr/include/stdio.h
/path/to/foo.c	source BUILDID /path/to/foo.c
\../bar/foo.c AT_comp_dir=/zoo/	source BUILDID /zoo//../bar/foo.c
.TE

.SS section \fIBUILDID\fP \fISECTION-NAME\fP

If the given buildid is known to the server, this request will result
in a binary file that contains the raw section contents, whether it is
found in an executable or a separated debuginfo file.


.SS metadata \fIKEY\fP \fIVALUE\fP

All designated debuginfod servers are queried for metadata about all
files that match a given key/value query in their index.  The results
include names and buildids, which may be used in future queries to
fetch actual files.

.TS
l l l .
KEY	VALUE	DESCRIPTION

\fBfile\fP	\fIpath\fP	exact match \fIpath\fP, including in archives
\fBglob\fP	\fIpattern\fP	shell-style glob match \fIpattern\fP, including in archives, as in fnmatch(FNM_PATHNAME)
\fBbuildid\fP	\fIhex\fP	all executables and debuginfo files matching the given buildid
.TE

The resulting output will look something like the following
{
  "results":[
    {
      "type":"executable",
      "buildid":"f0aa15b8aba4f3c28cac3c2a73801fefa644a9f2",
      "file":"/usr/local/bin/hello",
      "archive":"/opt/elfutils/tests/test-2290642/R/rhel7/hello2-1.0-2.x86_64.rpm"
    },
    {
      "type":"executable",
      "buildid":"bc1febfd03ca05e030f0d205f7659db29f8a4b30",
      "file":"hello2"
    }
  ],
  "complete":true
}'

The results of the search are output to \fBstdout\fP as a JSON object
containing an array of objects, supplying metadata about each match, as
well as a boolean value corresponding to the completeness of the result.
The result is considered complete if all of the queries to upstream servers
returned complete results and the local query succeeded. This metadata report
may be cached.  It may be incomplete and may contain duplicates.  
Additional JSON object fields may be present.

.TS
l l l .
NAME	TYPE	DESCRIPTION

\fBbuildid\fP	string	hexadecimal buildid associated with the file
\fBtype\fP	string	one of \fBdebuginfo\fP or \fBexecutable\fP
\fBfile\fP	string	matched file name, outside or inside the archive
\fBarchive\fP	string	archive containing matched file name, if any
.TE

It's worth noting that \fBtype\fP cannot be \fBsource\fP since in order
to perform such a search fast enough additional indexing would need to be added to
the database which would nearly double its size.

The search also always combines both files and archives in the results
and at this time further granularity is not availible.

.SH "OPTIONS"

.TP
.B "\-v" "\-\-verbose"
Increase verbosity, including printing frequent download-progress messages
and printing the http response headers from the server.

.TP
.B "\-\-help" "\-\-usage"
Print some help and exit.

.SH "SECURITY"

If IMA signature(s) are available from the RPMs that contain
requested files, then
.BR debuginfod
will extract those signatures into response headers, and
.BR debuginfod-find
will perform verification upon the files.
Validation policy is controlled via tags inserted into
$DEBUGINFOD_URLS.  By default, 
.BR debuginfod-find
acts in ignore mode.

If accessed across HTTP rather than HTTPS, the
network should be trustworthy.  Authentication information through
the internal \fIlibcurl\fP library is not currently enabled, except
for the basic plaintext \%\fIhttp[s]://userid:password@hostname/\fP style.
(The debuginfod server does not perform authentication, but a front-end
proxy server could.)

.nr zZ 1
.so man7/debuginfod-client-config.7

.SH "SEE ALSO"
.I "debuginfod(8)"
.I "debuginfod_find_debuginfod(3)"


CHANGELOG.md
# CHANGELOG

### v0.7.2 (2026-08-22)

- Bug-fix: due to the net-layer new signal handlers and julia's own gc changes,
net-layer's thread was causing the process to crash on julia 1.12.

### v0.7.1 (2026-08-21)

- New: `start_app` now can accept either a path to a script or a `Function` as
the app entry point.
- New: New function `inject_html` that allows users (that is, developers using
Magic.jl) to inject custom HTML into pages served to their clients.
- New: port 0 can now be passed to `start_app` to let the OS pick an available
port for the server.
- New: new boolean argument `init_and_quit` for `start_app`. If true,
`start_app` will only initialize the server and immediately return. Useful for
automated tests of dry-runs.
- Bug-fix: `download_button` was broken due to incorrect generation of the
file path in the backend.
- Bug-fix: font path on avatar example updated to handle cases where the .Magic
location is arbitratily defined.
- Bug-fix: when using relative path, `add_font` was not aways properly striping
away the .Magic/served-files path from the provided path.
- Internal: Introduces the first batch of tests to the package, some unit tests,
some end-to-end tests.
- Internal: New dependencies backward.hpp, elfutils, xz, bzip2 and zstd, used
for stacktrace printing on fatal error on, the net-layer on linux.
- Internal: New crash recovery mechanism on dev mode on linux.

### v0.7.0 (2026-08-17)

- New: Docstrings for every public function so that users can get help from a
Julia's REPL session.
- New: `get_dot_magic_dir()` and `get_dot_magic_path()`.
- Change: `.Magic` default location changed to be the process working directory.
More about `.Magic` in the "Basic Concepts" section of the documentation.
- Internal: Tests implemented for `start_app()` input arguments.
- Internal: Code refactoring. Moved code from `Magic.jl` to the newly created
source files `layout.jl`, `elements.jl` and `logic.jl`.

### v0.6.2 (2026-08-14)

- Bug-fix: Fragments function name generation was broken. Fixed.

### v0.6.1 (2026-02-14)

- New: Widget `slider()`.
- New: Sliding curves example.
- Bug-fix: on small screens, `row`s had overflow problems similar to what
columns used to have. It seems that the flex model (at least on chrome) works
with a default `min-width` based on "intrinsic width of child elements",
whatever that is. The fix is to enforce a `min-width` of `0` on our layout
elements.

### v0.6.0 (2026-02-08)

- Breaking-change: Removed argument `inner_func` from `set_page_layout()` and
thus the ability to have a `do-end` block with this function. If you want a
`do-end` block that does the same, implement it with `main_area()` instead,
after calling `set_page_layout()`.
- New: New widget `number_input()`.
- New: Example `probability.jl`.
- Bug-fix: Fixes `file_uploader()` bug where selecting a file, clearing the
selection and selecting the same file didn't work.
https://github.com/nidoro/Magic.jl/issues/8
- Bug-fix: Fixes `file_uploader()` bug where canceling the file dialog when a
file was selected triggered a change callback, when it shouldn't.
- Bug-fix: Dry-run logic was breaking trying to access session 0 after the
`handle_client_left(Cint(0))` has been called.

### v0.5.0 (2026-01-31)

- Breaking-change: renamed `gen_resource_path` to `gen_serveable_path`.
Additionally, this function now generates file names without extension when
an empty string is passed to extension, which is now the default value of the
argument.
- New: Widget `file_uploader()`.
- New: Widget `download_button()`.
- New: Function `make_serveable_copy()`.
- New: Function `move_to_serveable_dir()`.
- New: Example `image-filters.jl`.
- New: A `.gitignore` file is now automatically generated inside `.Magic` if
it doesn't exist.
- New: New `start_app()` arguments `upload_max_size` and `upload_max_files`.
- Change: if a non-serveable path is passed to `image()` and the path points
to an existing file, a serveable copy is automatically created using
`make_serveable_copy()`.
- Change: Dry-run errors do not kill the server anymore. Rather, the user can
see the error when accessing the web app, and work on the fix while using the
hot-reloading mechanism.
- Performance: The temporary module where the user script is evaluated is now
session-persistent rather then recreated on every rerun.
- Deprecated: `@once` macro is deprecated, as it turns out it is not needed now
that the app module is session-persistent.

### v0.4.2 (2026-01-22)

- Updates documentation installation instructions now that the package is
registered in the julia General registry.
- Bug-fix: Fixes catch block of rerun not reporting the error to the user.
https://github.com/nidoro/Magic.jl/issues/7
- Bug-fix: After introduction of the `docs_path` argument to `start_app()` on
v0.3.0, we didn't update the CLI arguments to mirror that change.

### v0.4.1 (2026-01-20)

- Hot-fix: Fixes server crash after session is closed
https://github.com/nidoro/Magic.jl/issues/6

### v0.4.0 (2026-01-18)

- Breaking-change: Package name change from Lit.jl to Magic.jl.
- New: Introduced `lifetime` argument to `gen_resource_path()`.

### v0.3.2 (2026-01-16)

- Hot-fix: examples and CHANGELOG.

### v0.3.1 (2026-01-16)

- Bug-fix: Unqueued rerun request are now validated as well.
- Bug-fix: Implemented invalid state acknoledgment system, to avoid this
situation:
    1. Client sends 2 events in a row. The second is invalid, given the first
    one. Invalid state is sent from server to client.
    2. Before receiving the invalid state message, the client sends a third
    event, which is valid. Then restores old state and unfades.
    3. The server processes the new event and returns the newer state.
    4. An update then happens in the client, which is unnexpected, because the
    return from transparency should mean that all events have been processed,
    and there are no other ones in queue.
- Cache-busting: Introduced cache-busting system. Generated HTML pages now use

net-layer/third/elfutils/debuginfod/debuginfod-find.c
/* Command-line frontend for retrieving ELF / DWARF / source files
   from the debuginfod.
   Copyright (C) 2019-2023 Red Hat, Inc.
   This file is part of elfutils.

   This file is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   elfutils is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */


#include "config.h"
#include "printversion.h"
#include "debuginfod.h"
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <argp.h>
#include <unistd.h>
#include <fcntl.h>
#include <gelf.h>
#include <libdwelf.h>
#include <signal.h>
#ifndef DUMMY_LIBDEBUGINFOD
#include <json-c/json.h>
#endif

/* Name and version of program.  */
ARGP_PROGRAM_VERSION_HOOK_DEF = print_version;

/* Bug report address.  */
ARGP_PROGRAM_BUG_ADDRESS_DEF = PACKAGE_BUGREPORT;

/* Short description of program.  */
static const char doc[] = N_("Request debuginfo-related content "
                             "from debuginfods listed in $" DEBUGINFOD_URLS_ENV_VAR ".");

/* Strings for arguments in help texts.  */
static const char args_doc[] = N_("debuginfo BUILDID\n"
                                  "debuginfo PATH\n"
                                  "executable BUILDID\n"
                                  "executable PATH\n"
                                  "source BUILDID /FILENAME\n"
                                  "source PATH /FILENAME\n"
                                  "section BUILDID SECTION-NAME\n"
                                  "section PATH SECTION-NAME\n"
                                  "metadata (glob|file|buildid|KEY) (GLOB|FILENAME|BUILDID|VALUE)\n"
                                  );

/* Definitions of arguments for argp functions.  */
static const struct argp_option options[] =
  {
   { "verbose", 'v', NULL, 0, "Increase verbosity.", 0 },
   { NULL, 0, NULL, 0, NULL, 0 }
  };

/* debuginfod connection handle.  */
static debuginfod_client *client;
static int verbose;
static volatile sig_atomic_t interrupted;
static bool saw_progress_envvar;

static void
handle_sigint(int signo __attribute__((__unused__)))
{
  interrupted = 1;
}

static int
progressfn(debuginfod_client *c, long a, long b)
{
  if (interrupted)
    return 1;
  if (verbose < 1)
    {
      /* Fallback to the default progressfn if verbose isn't set and
	 DEBUGINFOD_PROGRESS is set.  */
      if (saw_progress_envvar)
	return debuginfod_default_progressfn (c, a, b);
      return 0;
    }

  static bool first = true;
  static struct timespec last;
  struct timespec now;
  uint64_t delta;
  if (!first)
    {
      clock_gettime (CLOCK_MONOTONIC, &now);
      delta = ((now.tv_sec - last.tv_sec) * 1000000
	       + (now.tv_nsec - last.tv_nsec) / 1000);
    }
  else
    {
      first = false;
      delta = 250000;
    }

  /* Show progress the first time and then at most 5 times a second. */
  if (delta > 200000)
    {
      fprintf (stderr, "Progress %ld / %ld\n", a, b);
      clock_gettime (CLOCK_MONOTONIC, &last);
    }
  return 0;
}


static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
  (void) arg;
  (void) state;
  switch (key)
    {
    case 'v': verbose++;
      if (verbose > 1)
        debuginfod_set_verbose_fd (client, STDERR_FILENO);
      break;
    default: return ARGP_ERR_UNKNOWN;
    }
  return 0;
}


/* Data structure to communicate with argp functions.  */
static struct argp argp =
  {
   options, parse_opt, args_doc, doc, NULL, NULL, NULL
  };



int
main(int argc, char** argv)
{
  elf_version (EV_CURRENT);

  client = debuginfod_begin ();
  if (client == NULL)
    {
      fprintf(stderr, "Couldn't create debuginfod client context\n");
      return 1;
    }

  /* Set SIGINT handler and progressfn so that temp files can be cleaned
     up when a download is cancelled.  */
  struct sigaction sa;
  sigemptyset (&sa.sa_mask);
  sa.sa_flags = 0;
  sa.sa_handler = handle_sigint;
  sigaction (SIGINT, &sa, NULL);

  if (getenv(DEBUGINFOD_PROGRESS_ENV_VAR) != NULL)
    saw_progress_envvar = true;

  debuginfod_set_progressfn (client, & progressfn);

  /* Exercise user data pointer, to support testing only. */
  debuginfod_set_user_data (client, (void *)"Progress");

  int remaining;
  (void) argp_parse (&argp, argc, argv, ARGP_IN_ORDER|ARGP_NO_ARGS, &remaining, NULL);

  if (argc < 2 || remaining+1 >= argc) /* no arguments or at least two non-option words */
    {
      argp_help (&argp, stderr, ARGP_HELP_USAGE, argv[0]);
      return 1;
    }

  /* If we were passed an ELF file name in the BUILDID slot, look in there. */
  unsigned char* build_id = (unsigned char*) argv[remaining+1];
  int build_id_len = 0; /* assume text */
  Elf* elf = NULL;

  /* Process optional buildid given via ELF file name, for some query types only. */
  if (strcmp(argv[remaining], "debuginfo") == 0
      || strcmp(argv[remaining], "executable") == 0
      || strcmp(argv[remaining], "source") == 0
      || strcmp(argv[remaining], "section") == 0)
    {
      int any_non_hex = 0;
      int i;
      for (i = 0; build_id[i] != '\0'; i++)
        if ((build_id[i] >= '0' && build_id[i] <= '9') ||
            (build_id[i] >= 'a' && build_id[i] <= 'f'))
          ;
        else
          any_non_hex = 1;
      
      int fd = -1;
      if (any_non_hex) /* raw build-id */
        {
          fd = open ((char*) build_id, O_RDONLY);
          if (fd < 0)
            fprintf (stderr, "Cannot open %s: %s\n", build_id, strerror(errno));
        }
      if (fd >= 0)
        {
          elf = dwelf_elf_begin (fd);
          if (elf == NULL)
            fprintf (stderr, "Cannot open as ELF file %s: %s\n", build_id,
                     elf_errmsg (-1));
        }
      if (elf != NULL)
        {
          const void *extracted_build_id;
          ssize_t s = dwelf_elf_gnu_build_id(elf, &extracted_build_id);
          if (s > 0)
            {
              /* Success: replace the build_id pointer/len with the binary blob
                 that elfutils is keeping for us.  It'll remain valid until elf_end(). */
              build_id = (unsigned char*) extracted_build_id;
              build_id_len = s;
            }
          else
            fprintf (stderr, "Cannot extract build-id from %s: %s\n", build_id, elf_errmsg(-1));
        }
    }

  char *cache_name;
  int rc = 0;

  /* By default the stdout output is the path of the cached file.
     Some requests (ex. metadata query may instead choose to do a different output,
     in that case a stringified json object) */
  bool print_cached_file = true;
  /* Check whether FILETYPE is valid and call the appropriate
     debuginfod_find_* function. If FILETYPE is "source"
     then ensure a FILENAME was also supplied as an argument.  */
  if (strcmp(argv[remaining], "debuginfo") == 0)
    rc = debuginfod_find_debuginfo(client,
				   build_id, build_id_len,
				   &cache_name);
  else if (strcmp(argv[remaining], "executable") == 0)
    rc = debuginfod_find_executable(client,
                                    build_id, build_id_len,
				    &cache_name);
  else if (strcmp(argv[remaining], "source") == 0)
    {
      if (remaining+2 >= argc || argv[remaining+2][0] == '\0')
        {
          fprintf(stderr, "If FILETYPE is \"source\" then FILENAME must be given\n");
          return 1;
        }
      rc = debuginfod_find_source(client,
                                  build_id, build_id_len,
				  argv[remaining+2], &cache_name);
    }
  else if (strcmp(argv[remaining], "section") == 0)
    {
      if (remaining+2 >= argc)
	{
	  fprintf(stderr,
		  "If FILETYPE is \"section\" then a section name must be given\n");
	  return 1;
	}
      rc = debuginfod_find_section(client, build_id, build_id_len,
				   argv[remaining+2], &cache_name);
    }
  else if (strcmp(argv[remaining], "metadata") == 0) /* no buildid! */
    {
      if (remaining+2 == argc)
        {
          fprintf(stderr, "Require KEY and VALUE for \"metadata\"\n");
          return 1;
        }
      
      rc = debuginfod_find_metadata (client, argv[remaining+1], argv[remaining+2],
                                     &cache_name);
#ifndef DUMMY_LIBDEBUGINFOD
      if (rc >= 0)
        {
          /* We output a pprinted JSON object, not the regular debuginfod-find cached file path */
          print_cached_file = false;
          json_object *metadata = json_object_from_file(cache_name);
          if(metadata)
            {
              printf("%s\n", json_object_to_json_string_ext(metadata,
                                                            JSON_C_TO_STRING_PRETTY
#ifdef JSON_C_TO_STRING_NOSLASHESCAPE /* json-c 0.15 */
                                                            | JSON_C_TO_STRING_NOSLASHESCAPE
#endif
                                                            ));
              json_object_put(metadata);
            }
          else
            {
              fprintf(stderr, "%s does not contain a valid JSON format object\n", cache_name);
              return 1;
            }
        }
#endif
    }
  else
    {
      argp_help (&argp, stderr, ARGP_HELP_USAGE, argv[0]);
      return 1;
    }

  if (verbose)
    {
      const char* headers = debuginfod_get_headers(client);
      if (headers)
        fprintf(stderr, "Headers:\n%s", headers);
      const char* url = debuginfod_get_url (client);
      if (url != NULL)
        fprintf(stderr, "Downloaded from %s\n", url);
    }

  debuginfod_end (client);
  if (elf)
    elf_end(elf);

  if (rc < 0)
    {
      if (interrupted != 0)
        fputs ("Server query cancelled\n", stderr);
      else
        fprintf(stderr, "Server query failed: %s\n", strerror(-rc));

      return 1;
    }
  else
    close (rc);

  if(print_cached_file) printf("%s\n", cache_name);
  free (cache_name);

  return 0;
}
